controllers/update_status_test.go

```diff
@@ -41,6 +43,352 @@ import (
+				Expect(err).NotTo(HaveOccurred())
+
+				err := validateProcessGroup(context.TODO(), clusterReconciler, cluster, pod, pvc, pod.ObjectMeta.Annotations[fdbv1beta2.LastConfigMapKey], processGroupStatus, cluster.IsTaintFeatureDisabled(), logger)
+
+				Expect(err).NotTo(HaveOccurred())
+				Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(1))
+				Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected)).NotTo(Equal(nil))
+			})
+			It("should get both NodeTaintDetected and NodeTaintReplacing condition", func() {
+				// Taint the node
+				node.Spec.Taints = []corev1.Taint{
+					{
+						Key:       taintKeyMaintenance,
+						Value:     "rack_maintenance",
+						Effect:    corev1.TaintEffectNoExecute,
+						TimeAdded: &metav1.Time{Time: time.Now().Add(-time.Second * time.Duration(taintKeyMaintenanceDuration+1))},
+					},
+				}
+				log.Info("Taint node", "Node name", pod.Name, "Node taints", node.Spec.Taints, "Now", time.Now())
+				// Make taint in effect
+				err = k8sClient.Update(context.TODO(), node)
+				Expect(err).NotTo(HaveOccurred())
+
+				err := validateProcessGroup(context.TODO(), clusterReconciler, cluster, pod, pvc, pod.ObjectMeta.Annotations[fdbv1beta2.LastConfigMapKey], processGroupStatus, cluster.IsTaintFeatureDisabled(), logger)
+
+				Expect(err).NotTo(HaveOccurred())
+				Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(2))
+				Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected)).NotTo(Equal(nil))
+				Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintReplacing)).NotTo(Equal(nil))
+			})
+			It("should disable taint feature with configured negative-duration wildcard taint key ", func() {
+				// Taint the node
+				node.Spec.Taints = []corev1.Taint{
+					{
+						Key:       taintKeyMaintenance,
+						Value:     "rack_maintenance",
+						Effect:    corev1.TaintEffectNoExecute,
+						TimeAdded: &metav1.Time{Time: time.Now().Add(-time.Second * time.Duration(taintKeyMaintenanceDuration+1))},
+					},
+				}
+				err = k8sClient.Update(context.TODO(), node)
+				Expect(err).NotTo(HaveOccurred())
+				log.Info("Taint node", "Node name", pod.Name, "Node taints", node.Spec.Taints, "Now", time.Now())
+
+				// Disable cluster's taint mechanism
+				cluster.Spec.AutomationOptions.Replacements.TaintReplacementOptions = []fdbv1beta2.TaintReplacementOption{
+					{
+						Key:               &taintKeyWildcard,
+						DurationInSeconds: &taintKeyWildcardNegativeDuration,
+					},
+					{
+						Key:               &taintKeyMaintenance,
+						DurationInSeconds: &taintKeyMaintenanceDuration,
+					},
+				}
+
+				err := validateProcessGroup(context.TODO(), clusterReconciler, cluster, pod, pvc, pod.ObjectMeta.Annotations[fdbv1beta2.LastConfigMapKey], processGroupStatus, cluster.IsTaintFeatureDisabled(), logger)
+
+				Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(0))
+				Expect(err).NotTo(HaveOccurred())
+			})
```

Comment:
```suggestion
			When("the taint was just added long enough to to mark the Pod for replacement", func() {
				BeforeEach(func() {
					// Taint the node
					node.Spec.Taints = []corev1.Taint{
						{
							Key:       taintKeyMaintenance,
							Value:     "rack_maintenance",
							Effect:    corev1.TaintEffectNoExecute,
							TimeAdded: &metav1.Time{Time: time.Now().Add(-time.Second * time.Duration(taintKeyMaintenanceDuration+1))},
						},
					}
				})

				It("should get both NodeTaintDetected and NodeTaintReplacing condition", func() {
					Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(2))
					Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected)).NotTo(Equal(nil))
					Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintReplacing)).NotTo(Equal(nil))
				})

				When("the wildcard taint has a negative duration", func() {
					BeforeEach(func() {
						// Disable cluster's taint mechanism
						cluster.Spec.AutomationOptions.Replacements.TaintReplacementOptions = []fdbv1beta2.TaintReplacementOption{
							{
								Key:               &taintKeyWildcard,
								DurationInSeconds: &taintKeyWildcardNegativeDuration,
							},
							{
								Key:               &taintKeyMaintenance,
								DurationInSeconds: &taintKeyMaintenanceDuration,
							},
						}
					})

					It("should disable taint feature ", func() {
						Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(0))
						Expect(err).NotTo(HaveOccurred())
					})
				})
			})
```

We should restructure the test case like this (with nest statements) to make it easier to understand the test cases and the nested `When` statements should make it easier to understand the actual test case we are verifying.